### PR TITLE
Dolfyn updates

### PR DESCRIPTION
**Pull request by jmcvey3**
A number of updates:
1. Fixes for TRDI reader: #182, #206, #207
2. Fix for VMDAS files with interleaved pings
3. Motion correction for duty-cycled ADVs
4. lncorporated logging module for into reader debuggers
5. Ensure datatypes are saved as float32 instead of float64 to reduce filesize
6. Added another ADCP clean function
7. Updated docstrings to mhkit formatting
8. Other minor bugfixes

### CODE REVIEW

**Comment by ssolson** [comment]
James thank you for submitting the PR. To start the review I have a couple of questions about large log files:

1. 1.16 MB `examples/data/dolfyn/test_data/AWAC_test01.log`
2. 6.74 MB `examples/data/dolfyn/test_data/vector_data_imu01.log`
3. 3.05 MB `examples/data/dolfyn/test_data/RDI_withBT.log`

These add about 11 MB of log data. Is there a way we could reduce these log file sizes such as checking against a smaller dataset (and therefore a smaller associated log file), checking only a portion of the produced log, or mocking the data somehow?

File: mhkit/dolfyn/adv/motion.py [hunk 1]
```diff
@@ -86,16 +87,55 @@ def _set_accel(self, ):
...
+            self.ds.time[n_burst:].values - self.ds.time[:-n_burst].values)/1e9
+
+        rng = actual_interval.max() - actual_interval.min()
+        mean = actual_interval.mean()
+        # Range will vary depending on how datetime64 rounds the timestamp
+        # But isn't an issue if it does
+        if rng > 2 or (mean > interval+1 and mean < interval-1):
+            raise Exception("Bad duty cycle detected")
+
+        # If this passes, it means we're save to blindly skip n_burst for every integral
```
Comment (ssolson) [inline mhkit/dolfyn/adv/motion.py:112]: I think it should be "safe" instead of save

**Review by ssolson — COMMENTED** [review]
A couple of comments for consideration. Still working my way through these so this is just what I have so far.

**Comment by jmcvey3** [comment]
> James thank you for submitting the PR. To start the review I have a couple of questions about large log files:
> 
> 1. 1.16 MB `examples/data/dolfyn/test_data/AWAC_test01.log`
> 2. 6.74 MB `examples/data/dolfyn/test_data/vector_data_imu01.log`
> 3. 3.05 MB `examples/data/dolfyn/test_data/RDI_withBT.log`
> 
> These add about 11 MB of log data. Is there a way we could reduce these log file sizes such as checking against a smaller dataset (and therefore a smaller associated log file), checking only a portion of the produced log, or mocking the data somehow?

Good point, yes I can reduce the size of these.

File: mhkit/dolfyn/adp/clean.py [hunk 2]
```diff
@@ -170,19 +171,24 @@ def nan_beyond_surface(ds, val=np.nan):
       The adcp dataset to clean
     val : nan or numeric
       Specifies the value to set the bad values to (default np.nan).
+    inplace : bool (default: False)
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/adp/clean.py:174]: It looks like it puts everything after the first colon in italics (https://mhkit-software.github.io/MHKiT/mhkit-python/api.dolfyn.html#io)

File: mhkit/dolfyn/adv/clean.py [hunk 3]
```diff
@@ -2,14 +2,15 @@
...
+from ..velocity import VelBinner
 from ..tools.misc import group, slice1d_along_axis
 warnings.filterwarnings('ignore', category=np.RankWarning)
 
 sin = np.sin
 cos = np.cos
 
 
-def clean_fill(u, mask, npt=12, method='cubic', max_gap=6):
+def clean_fill(u, mask, npt=12, method='cubic', maxgap=6):
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/adv/clean.py:13]: Good catch

File: mhkit/dolfyn/io/rdi_lib.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+    # def checksum(self,):
+    #     """
+    #     The next byte(s) are the expected checksum.  Perform the checksum.
+    #     """
+    #     if self.cs:
+    #         cs = self.read(1, self.cs._frmt)
+    #         self.cs(cs, True)
+    #     else:
+    #         raise Exception('CheckSum not requested for this file')
```
Comment (ssolson) [inline mhkit/dolfyn/io/rdi_lib.py:59]: Is this something you plan to bring back in the future?

File: mhkit/dolfyn/io/rdi_lib.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+    def read_ui32(self, n):
+        return self.read(n, 'L')
+
+    def read_i32(self, n):
+        return self.read(n, 'l')
+
+# ics = 0  # This is a holder for the checksum index
+# class checksum():
+#     # Checksum for TRDI
```
Comment (ssolson) [inline mhkit/dolfyn/io/rdi_lib.py:120]: Similar to above is a checksum() class going to be added in a future update?

File: mhkit/dolfyn/io/rdi.py [hunk 6]
```diff
@@ -630,23 +780,53 @@ def read_cfgseg(self,):
...
+            cfg['beam_angle'] = fd.read_ui8(1)
+            self._nbyte += 6
+        # # Documented in latest PDDecoder 10/12/22 but must be instrument specific
+        # if cfg['prog_ver'] >= 56:
+        #     fd.seek(1, 1)
+        #     pings_per_ensemble = fd.read_ui16(1)
+        #     exact_freq = fd.read_ui8(3)
+        #     #cfg['exact_freq'] = int("".join(str(x) for x in exact_freq))
+        #     self._nbyte += 6
+
```
Comment (ssolson) [inline mhkit/dolfyn/io/rdi.py:817]: Looks like this was replaced by the definition above. Do we want to keep this?

**Review by ssolson — COMMENTED** [review]
@jmcvey3 a couple of minor comments. Great job on getting the docstring consistent. 

Is there anything I missed that you would like to talk through before we approve the PR?

File: mhkit/dolfyn/adv/motion.py [hunk 7]
```diff
@@ -15,25 +16,23 @@ def _get_body2imu(make_model):
...
+      The frequency at which to high-pass filter the acceleration
+      sigal to remove low-frequency drift. Default = 0.03 Hz
     vel_filtfreq : float (optional)
       a second frequency to high-pass filter the integrated
-      acceleration.  (default: 1/3 of accel_filtfreq)
-
+      acceleration.  Default = 1/3 of `accel_filtfreq`
     """
+
     _default_accel_filtfreq = 0.03
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/adv/motion.py:36]: No, there's two filter frequencies here. One for acceleration and one for velocity. The velocity filter frequency defaults to 1/3 of the acceleration frequency (0.03), so it's default is 0.01.

File: mhkit/dolfyn/adv/turbulence.py [hunk 8]
```diff
@@ -259,12 +221,12 @@ def dissipation_rate_LT83(self, psd, U_mag, f_range=[6.28, 12.57]):
         out = (psd.isel(freq=idx) *
                freq.isel(freq=idx)**(5/3) / a).mean(axis=-1)**(3/2) / U
 
-        out = xr.DataArray(out, name='dissipation_rate',
+        out = xr.DataArray(out.astype('float32'),
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/adv/turbulence.py:224]: Yes, turns out the name is immediately overwritten once a dataArray is assigned a variable name.

File: mhkit/dolfyn/io/base.py [hunk 9]
```diff
@@ -25,6 +25,7 @@ def _get_filetype(fname):
     with open(fname, 'rb') as rdr:
         bytes = rdr.read(40)
     code = bytes[:2].hex()
+    #print("{} - {}".format(fname.rsplit('/')[-1], bytes))
```
[the remark this answers is not in the document]
Comment (jmcvey3) [reply mhkit/dolfyn/io/base.py:28]: Thanks, yes it is

File: mhkit/dolfyn/io/rdi_lib.py (hunk 5 above)
Comment (jmcvey3) [reply mhkit/dolfyn/io/rdi_lib.py:120]: No this is old. Looks like I forgot to drop it

File: mhkit/dolfyn/io/rdi.py (hunk 6 above)
Comment (jmcvey3) [reply mhkit/dolfyn/io/rdi.py:817]: I'll remove it for now from the main branch until it's needed. I'm not entirely sure why it's in their documentation.

**Comment by jmcvey3** [comment]
@ssolson In case I hadn't mentioned it, I think this PR should now be covered. I'll need to update the docs, but I believe it's easiest to wait until this is merged into Develop?

**Comment by ssolson** [comment]
Thanks James. I'm going to merge into Develop. Please update the Docs as you mentioned.

**Review by ssolson — APPROVED** [review]

**Pull request merged**